File: yunva17/Week_19/BOJ_21278_호석이두마리치킨/BOJ_21278_호석이두마리치킨.js
let input = require('fs')
  // .readFileSync('/dev/stdin')
  .readFileSync('Week_19/BOJ_21278_호석이두마리치킨/input.txt')
  .toString()
  .trim()
  .split('\n');

const [N, M] = input[0].split(' ').map(Number);

const arr = input.slice(1).map((v) => v.split(' ').map(Number));

const INF = Infinity;

// 그래프 초기화
const graph = Array.from(Array(N + 1), () => Array(N + 1).fill(INF));
arr.forEach(([a, b]) => {
  graph[a].push(b);
  graph[b].push(a);
});

// 자기 자신 거리 0
for (let i = 1; i <= N; i++) {
  graph[i][i] = 0;
}

// 도로 정보 입력
arr.forEach(([a, b]) => {
  graph[a][b] = 1;
  graph[b][a] = 1;
});

// 플로이드 워셜
for (let k = 1; k <= N; k++) {
  for (let i = 1; i <= N; i++) {
    for (let j = 1; j <= N; j++) {
      if (graph[i][j] > graph[i][k] + graph[k][j]) {
        graph[i][j] = graph[i][k] + graph[k][j];
      }
    }
  }
}

// 가능한 치킨집의 조합 중 최소 비용 계산
let minSum = INF;
let answer = [];

const getSum = (a, b) => {
  let sum = 0;

  for (let i = 1; i <= N; i++) {
    if (i !== a && i !== b) {
      sum += Math.min(graph[i][a], graph[i][b]) * 2;
    }
  }

  if (sum < minSum) {
    minSum = sum;
    answer = [a, b];
  }
};

const check = () => {
  for (let i = 1; i <= N; i++) {
    for (let j = i + 1; j <= N; j++) {
      getSum(i, j);
    }
  }
};

check();

console.log(answer.join(' '), minSum);

File: violetadieu/Week_19/BOJ_21278_호석이두마리치킨.py
import sys
from typing import List

n, m = map(int, input().split())

graph = [[sys.maxsize] * (n + 1) for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, input().split())
    graph[a][b] = 1
    graph[b][a] = 1

for i in range(1, n + 1):
    graph[i][i] = 0

for k in range(1, n + 1):
    for a in range(1, n + 1):
        for b in range(1, n + 1):
            graph[a][b] = min(graph[a][b], graph[a][k] + graph[k][b])

min_sum = sys.maxsize
result: List[int] = []
for i in range(1, n):
    for j in range(2, n + 1):
        sum_val = 0
        for k in range(1, n + 1):
            sum_val += min(graph[k][i], graph[k][j]) * 2
        if sum_val < min_sum:
            min_sum = sum_val
            result = [i, j, sum_val]

print(result[0], result[1], result[2])


